src/ripple/app/paths/impl/PaySteps.cpp

```diff
@@ -0,0 +1,533 @@
+            {
+                JLOG (j.trace) << "Inserting implied account";
+                auto msr = make_DirectStepI (ctx(), cur->getAccountID (),
+                    curIssue.account, curIssue.currency);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+                Currency dummy;
+                impliedPE.emplace (STPathElement::typeAccount,
+                    curIssue.account, dummy, curIssue.account);
+                cur = &*impliedPE;
+            }
+        }
+        else if (isAccount (cur) && isOffer(next))
+        {
+            if (curIssue.account != cur->getAccountID ())
+            {
+                JLOG (j.trace) << "Inserting implied account before offer";
+                auto msr = make_DirectStepI (ctx(), cur->getAccountID (),
+                    curIssue.account, curIssue.currency);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+                Currency dummy;
+                impliedPE.emplace (STPathElement::typeAccount,
+                    curIssue.account, dummy, curIssue.account);
+                cur = &*impliedPE;
+            }
+        }
+        else if (isOffer (cur) && isAccount (next))
+        {
+            if (curIssue.account != next->getAccountID () &&
+                !isXRP (next->getAccountID ()))
+            {
+                JLOG (j.trace) << "Inserting implied account after offer";
+                auto msr = make_DirectStepI (ctx(), curIssue.account,
+                    next->getAccountID (), curIssue.currency);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            continue;
+        }
+
+        if (!isOffer (next) &&
+            hasCurrency(next) && next->getCurrency () != curIssue.currency)
+        {
+            auto const& nextCurrency = next->getCurrency ();
+            auto const& nextIssuer =
+                hasIssuer (next) ? next->getIssuerID () : curIssue.account;
+
+            if (isXRP (curIssue.currency))
+            {
+                JLOG (j.trace) << "Inserting implied XI offer";
+                auto msr = make_BookStepXI (
+                    ctx(), {nextCurrency, nextIssuer});
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            else if (isXRP (nextCurrency))
+            {
+                JLOG (j.trace) << "Inserting implied IX offer";
+                auto msr = make_BookStepIX (ctx(), curIssue);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
```

Comment:
I did a study on this and there is no difference between `push_back` and `emplace_back` when the single argument is the same type as the `vector::value_type`:  http://htmlpreview.github.io/?https://github.com/HowardHinnant/papers/blob/master/insert_vs_emplace.html  `insert` vs `emplace` is far more complicated though.